File: 2023/day02.cpp
// https://adventofcode.com/2023/day/2

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <print>
#include <ranges>
#include <span>

using Game = Vec<String>;

auto parse(String const& filename) -> Vec<Game> {
  return aoc::views::read_lines(filename) |
         stdv::transform([](str line) {
           let[_, game_str] = aoc::split_once(line, ':');
           return aoc::split_to_vec<String>(game_str, ';');
         }) |
         aoc::ranges::to<Vec<Game>>();
}

struct cube_config_t {
  int red{0};
  int green{0};
  int blue{0};

  bool operator==(const cube_config_t&) const = default;

  constexpr bool contains(const cube_config_t& other) const {
    return (red >= other.red) && (green >= other.green) && (blue >= other.blue);
  }

  friend constexpr cube_config_t max(const cube_config_t& lhs,
                                     const cube_config_t& rhs) {
    return {std::max(lhs.red, rhs.red), std::max(lhs.green, rhs.green),
            std::max(lhs.blue, rhs.blue)};
  }

  constexpr int power() const { return red * green * blue; }
};

inline constexpr cube_config_t config1{12, 13, 14};
inline constexpr cube_config_t config2{0, 0, 0};

/// Returns 1 if game possible, 0 if not (part 1)
/// Returns > 1 if config == {0,0,0} (part 2)
template <cube_config_t config>
fn cube_power(std::span<const str> games) -> i32 {
  static constexpr bool part2 = (config == config2);
  cube_config_t max_config{1, 1, 1};
  for (let game_str : games) {
    auto cubes_str = aoc::split_to_array<3>(game_str, ',');
    cube_config_t current_config;
    for (let cube_str : cubes_str) {
      if (cube_str.empty()) {
        continue;
      }
      let[number_str, color] = aoc::split_once(aoc::trim(cube_str), ' ');
      let number = aoc::to_number<i32>(number_str);
      if (color == "red") {
        current_config.red = number;
      } else if (color == "green") {
        current_config.green = number;
      } else if (color == "blue") {
        current_config.blue = number;
      } else {
        AOC_UNREACHABLE("Invalid color");
      }
      if constexpr (!part2) {
        if (!config.contains(current_config)) {
          return 0;
        }
      } else {
        max_config = max(max_config, current_config);
      }
    }
  }
  return max_config.power();
}

template <cube_config_t config>
fn solve_case(Vec<Game> const& games) -> i32 {
  auto sum = i32{};
  for (let[id, rounds] : games | stdv::enumerate) {
    let rounds_sv = rounds |
                    aoc::views::transform_cast<str>() |
                    aoc::ranges::to<Vec<str>>();
    auto power = cube_power<config>(std::span<const str>{rounds_sv});
    if constexpr (config != config2) {
      power *= id + 1;
    }
    sum += power;
  }
  return sum;
}

static_assert(1 ==
              cube_power<config1>(std::span<const str>{
                  {"3 blue, 4 red", "1 red, 2 green, 6 blue", "2 green"}}));
static_assert(0 == cube_power<config1>(std::span<const str>{
                       {"8 green, 6 blue, 20 red", "5 blue, 4 red, 13 green",
                        "5 green, 1 red"}}));
static_assert(48 ==
              cube_power<config2>(std::span<const str>{
                  {"3 blue, 4 red", "1 red, 2 green, 6 blue", "2 green"}}));
static_assert(1560 == cube_power<config2>(std::span<const str>{
                          {"8 green, 6 blue, 20 red", "5 blue, 4 red, 13 green",
                           "5 green, 1 red"}}));

int main() {
  std::println("Part 1");
  let example = parse("day02.example");
  AOC_EXPECT_RESULT(8, solve_case<config1>(example));
  let input = parse("day02.input");
  AOC_EXPECT_RESULT(2600, solve_case<config1>(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(2286, solve_case<config2>(example));
  AOC_EXPECT_RESULT(86036, solve_case<config2>(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2023/src/day02.rs
fn parse(filename: &str) -> Vec<Vec<String>> {
    aoc::file::read_lines(filename)
        .into_iter()
        .map(|line| {
            let (_, game_str) = line.split_once(':').unwrap();
            game_str.split(';').map(str::to_string).collect()
        })
        .collect()
}

#[derive(Clone, Copy, PartialEq)]
struct CubeConfig {
    red: i32,
    green: i32,
    blue: i32,
}
impl CubeConfig {
    fn contains(self, other: CubeConfig) -> bool {
        self.red >= other.red && self.green >= other.green && self.blue >= other.blue
    }

    fn max(self, other: CubeConfig) -> CubeConfig {
        CubeConfig {
            red: self.red.max(other.red),
            green: self.green.max(other.green),
            blue: self.blue.max(other.blue),
        }
    }

    fn power(self) -> i32 {
        self.red * self.green * self.blue
    }
}

const CONFIG1: CubeConfig = CubeConfig {
    red: 12,
    green: 13,
    blue: 14,
};

/// Returns 1 if game possible, 0 if not (part 1)
/// Returns > 1 if config == {0,0,0} (part 2)
fn cube_power<S: AsRef<str>>(config: Option<CubeConfig>, rounds: &[S]) -> i32 {
    let mut max_config = CubeConfig {
        red: 1,
        green: 1,
        blue: 1,
    };
    for round in rounds {
        let mut current = CubeConfig {
            red: 0,
            green: 0,
            blue: 0,
        };
        for cube_str in round.as_ref().split(',') {
            let cube_str = cube_str.trim();
            if cube_str.is_empty() {
                continue;
            }
            let (number_str, color) = cube_str.split_once(' ').unwrap();
            let number: i32 = number_str.parse().unwrap();
            match color {
                "red" => current.red = number,
                "green" => current.green = number,
                "blue" => current.blue = number,
                _ => unreachable!("Invalid color"),
            }
            if let Some(cfg) = config {
                if !cfg.contains(current) {
                    return 0;
                }
            } else {
                max_config = max_config.max(current);
            }
        }
    }
    max_config.power()
}

fn solve_case1(games: &[Vec<String>]) -> i32 {
    games
        .iter()
        .enumerate()
        .map(|(id, rounds)| cube_power(Some(CONFIG1), rounds) * (id as i32 + 1))
        .sum()
}

fn solve_case2(games: &[Vec<String>]) -> i32 {
    games.iter().map(|rounds| cube_power(None, rounds)).sum()
}

fn main() {
    println!("Asserts");
    aoc::expect_result!(
        1,
        cube_power(
            Some(CONFIG1),
            &["3 blue, 4 red", "1 red, 2 green, 6 blue", "2 green"]
        )
    );
    aoc::expect_result!(
        0,
        cube_power(
            Some(CONFIG1),
            &[
                "8 green, 6 blue, 20 red",
                "5 blue, 4 red, 13 green",
                "5 green, 1 red"
            ]
        )
    );
    aoc::expect_result!(
        48,
        cube_power(
            None,
            &["3 blue, 4 red", "1 red, 2 green, 6 blue", "2 green"]
        )
    );
    aoc::expect_result!(
        1560,
        cube_power(
            None,
            &[
                "8 green, 6 blue, 20 red",
                "5 blue, 4 red, 13 green",
                "5 green, 1 red"
            ]
        )
    );

    println!("Part 1");
    let example = parse("day02.example");
    aoc::expect_result!(8, solve_case1(&example));
    let input = parse("day02.input");
    aoc::expect_result!(2600, solve_case1(&input));

    println!("Part 2");
    aoc::expect_result!(2286, solve_case2(&example));
    aoc::expect_result!(86036, solve_case2(&input));
}
